pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -308,12 +318,39 @@ func (m *KV) GetListeningPort() int {
 	return int(m.memberlist.LocalNode().Port)
 }
 
-// JoinMembers joins the cluster with given members.
+// JoinMembers joins the cluster with given members. If the first attempt fails (e.g. no members reached) and
+// immediate abort is not set JoinMembers will retry with backoff up to max retries configured.
 // See https://godoc.org/github.com/hashicorp/memberlist#Memberlist.Join
 func (m *KV) JoinMembers(members []string) (int, error) {
 	return m.memberlist.Join(members)
 }
 
+func (m *KV) retryJoinWithBackoff(members []string) {
+	cfg := util.BackoffConfig{
+		MinBackoff: m.cfg.MinJoinBackoff,
+		MaxBackoff: m.cfg.MaxJoinBackoff,
+		MaxRetries: m.cfg.MaxJoinRetries,
+	}
+
+	backoff := util.NewBackoff(context.TODO(), cfg)
+
+	for backoff.Ongoing() {
+		reached, _ := m.memberlist.Join(members)
+		if reached == 0 {
+			level.Debug(util.Logger).Log("msg", "retry to join memberlist cluster", "node", m.cfg.NodeName, "attempts", backoff.NumRetries())
```

Comment:
Perhaps we should include error in the log? I know they can be quite verbose, but it's better than hiding it, especially when you need to troubleshoot this.